### PR TITLE
Chore/84 create tf chart

### PR DESCRIPTION
**Pull request by joshgamache**
Chart created for #84, [Project link](https://github.com/orgs/bcgov/projects/123/views/2?pane=issue&itemId=52301584). This is a chart-packaged version of the templates we've used to replace Terraform cloud (which was operating through *`cas-shelf`*). 

## Changes 🚧

- Repackaged templates into a single chart.
- Bundled terraform modules into chart directory.
  - We use Terraform identically in each project, no use in exposing this as config at this point. Simplifies the API for it greatly.
- Added automation to release the chart as a Github release.
- Added documentation for usage.

### CODE REVIEW

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+name: Release Charts
+
+on:
+  push:
+    branches:
+      - main
+      - chore/84-create-tf-chart
```
Comment (pbastia) [inline .github/workflows/release.yaml:7]: This can go

```suggestion

```

File: .github/workflows/release.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+jobs:
+  release:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
```
Comment (pbastia) [inline .github/workflows/release.yaml:14]:
```suggestion
        uses: actions/checkout@v4
```

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        with:
+          fetch-depth: 0
+
+      - name: Configure Git
+        run: |
+          git config user.name "$GITHUB_ACTOR"
+          git config user.email "$GITHUB_ACTOR@users.noreply.github.com"
+      - name: Install Helm
+        uses: azure/setup-helm@v1
```
Comment (pbastia) [inline .github/workflows/release.yaml:23]:
```suggestion
        uses: azure/setup-helm@v4
```

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        with:
+          version: v3.6.2
+
+      - name: Run chart-releaser
+        uses: helm/chart-releaser-action@v1.2.1
```
Comment (pbastia) [inline .github/workflows/release.yaml:28]:
```suggestion
        uses: helm/chart-releaser-action@v1.6.0
```

File: README.md [hunk 5]
```diff
@@ -57,7 +57,31 @@ Prior to using Helm to deploy applications to the OpenShift cluster, the CAS tea
...
+
+#### `~/terraform/modules/provision-storage-and-accounts`
+
+In tandem with the Helm chart is a Terraform module that creates GCP storage buckets, service accounts to access those buckets (admins and viewers) and injects those credentials into OpenShift for usage. These modules are pulled in via a configMap which pulls all files from the *parent chart's* `/terraform` directory.
+
+### Usage
+
+1. Import the Helm Chart into your project's main chart as a dependency.
+2. Add a subdirectory to the project chart's directory called `/terraform`. This is where you will place your `main.tf` and `variables.tf` files (and any other Terraform modules that you want imported and applied)
+3. Add a template into your chart with `{{ include "terraform-job.moduleImportMap" . }}`. This will use the partial from the terraform-job chart to pull in the terraform files from the project chart's context.
```
Comment (pbastia) [inline README.md:74]: We won't need these steps for v1, the main and variables files will be bundled.

File: helm/terraform-job/templates/_terraform-modules.yaml
Comment (pbastia) [inline helm/terraform-job/templates/_terraform-modules.yaml:1]: no underscore in name

File: helm/terraform-job/templates/terraform-apply.yaml [hunk 6]
```diff
@@ -0,0 +1,77 @@
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: terraform-apply
+  labels:
+    component: infrastructure
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    "helm.sh/hook": pre-install, pre-upgrade
```
Comment (pbastia) [inline helm/terraform-job/templates/terraform-apply.yaml:9]: `helm.sh/hook-weight: "10"`

File: helm/terraform-job/templates/terraform-role.yaml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: {{ .Values.serviceAccount.roleName }}
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    "helm.sh/hook": pre-install, pre-upgrade
```
Comment (pbastia) [inline helm/terraform-job/templates/terraform-role.yaml:7]: `helm.sh/hook-weight: "-10"`

File: helm/terraform-job/templates/terraform-service-account.yaml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: {{ .Values.serviceAccount.name }}
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    "helm.sh/hook": pre-install, pre-upgrade
```
Comment (pbastia) [inline helm/terraform-job/templates/terraform-service-account.yaml:7]: `helm.sh/hook-weight: "-10"`

File: helm/terraform-job/templates/terraform-role-binding.yaml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: "{{ .Values.serviceAccount.name }}-secret-admin-binding"
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    "helm.sh/hook": pre-install, pre-upgrade
```
Comment (pbastia) [inline helm/terraform-job/templates/terraform-role-binding.yaml:7]: `helm.sh/hook-weight: "-5"`

needs to be created after Role and SA

File: helm/terraform-job/values.yaml [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    cpu: 1000m
+    memory: 512Mi
+    ephemeral-storage: 1Gi
+  requests:
+    cpu: 100m
+    memory: 64Mi
+    ephemeral-storage: 256Mi
+
+terraform:
+  namespace_apps: '[]'
```
Comment (pbastia) [inline helm/terraform-job/values.yaml:18]:
```suggestion
  namespace_apps: ~
```

File: terraform/modules/provision-storage-and-accounts/main.tf
Comment (pbastia) [inline terraform/modules/provision-storage-and-accounts/main.tf:1]: needs to go to `/helm/terraform-job/files/...`

**Review by pbastia — CHANGES_REQUESTED** [review]
Amazing.
Rename chart to "terraform-bucket-provision"

File: .github/workflows/release.yaml (hunk 2 above)
Comment (joshgamache) [reply .github/workflows/release.yaml]: Updated in 8c2f4b6.

File: .github/workflows/release.yaml (hunk 3 above)
Comment (joshgamache) [reply .github/workflows/release.yaml]: Updated in 8c2f4b6.

File: .github/workflows/release.yaml (hunk 4 above)
Comment (joshgamache) [reply .github/workflows/release.yaml]: Updated in 8c2f4b6.

File: README.md (hunk 5 above)
Comment (joshgamache) [reply README.md]: Rewrote some docs in 2286add to reflect the bundling.

File: helm/terraform-job/templates/terraform-apply.yaml (hunk 6 above)
Comment (joshgamache) [reply helm/terraform-job/templates/terraform-apply.yaml]: Made sure all the templates have the hook weight ordering with 49bfe48.

File: helm/terraform-job/templates/terraform-role-binding.yaml (hunk 9 above)
Comment (joshgamache) [reply helm/terraform-job/templates/terraform-role-binding.yaml]: Made sure all the templates have the hook weight ordering with 49bfe48.

File: helm/terraform-job/templates/terraform-service-account.yaml (hunk 8 above)
Comment (joshgamache) [reply helm/terraform-job/templates/terraform-service-account.yaml]: Made sure all the templates have the hook weight ordering with 49bfe48.

File: helm/terraform-job/values.yaml (hunk 10 above)
Comment (joshgamache) [reply helm/terraform-job/values.yaml]: Revised in 25700c6.

File: terraform/modules/provision-storage-and-accounts/main.tf
Comment (joshgamache) [reply terraform/modules/provision-storage-and-accounts/main.tf:1]: Changed in ac3ab31c89a8c2e5a766d6163d206f3f434573c7.

File: terraform/modules/provision-storage-and-accounts/variables.tf
[the remark this answers is not in the document]
Comment (joshgamache) [reply terraform/modules/provision-storage-and-accounts/variables.tf:1]: Changed in ac3ab31c89a8c2e5a766d6163d206f3f434573c7.

File: helm/terraform-job/templates/terraform-role.yaml (hunk 7 above)
Comment (joshgamache) [reply helm/terraform-job/templates/terraform-role.yaml]: Made sure all the templates have the hook weight ordering with 49bfe48.

File: .github/workflows/release.yaml (hunk 1 above)
Comment (joshgamache) [reply .github/workflows/release.yaml]: dba3acb and done

**Pull request merged**